### PR TITLE
Improved plot updating logic in Muon Analysis

### LINKED ISSUE #25772: Muon GUI Usability improvements
This is an umbrella issue to keep track of minor improvements to the Muon GUI.

  * Make the workspace selector in the fitting tab searchable

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**

This updates the automatic plotting on the Muon Analysis GUI namely:
  * The plots will not rescale when they are updated
  * Only the latest fit is plotted and remains plotted until a new fit is done
  * The fit lines on the graph are correctly labelled as fit or diff.

**To test:**

  * You will need Access to the ISIS data archive to test this PR.
  * Open the Muon Analysis 2 interface and load some data (MUSR62260 tends to be a good example)
  * Go to the fitting tab add the GaussOsc function to the function browser (it's in the Muon subsection)
  * Click Fit
  * Check that the fitted line and the Diff appear on the graph. 
  * Change the frequency to 1 and try fitting again the lines on the graph should be updated to the new better fit.
  * Load in some new data, check that the plotted fit remains unchanged but that the plotted data is updated.

Fixes #25772 

*This does not require release notes* because this is an update to a new GUI

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_model.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from __future__ import (absolute_import, division, print_function)
 from mantidqt.plotting.functions import plot
 from mantid.api import AnalysisDataService
+import numpy as np
```
Comment (TTitcombe) [inline scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_model.py]: can we move this between `__future__` and `mantidqt.plotting`

File: scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_model.py [hunk 2]
```diff
@@ -35,21 +38,27 @@ def plot(self, workspace_list, title):
...
+            self.plot_figure.gca().set_xlim(left=0.0, right=15.0)
+            self.autoscale_y_to_data_in_view()
 
         self.plot_figure.canvas.set_window_title('Muon Analysis')
         self.plot_figure.gca().set_title(title)
 
         self.plot_figure.canvas.window().closing.connect(self._close_plot)
 
+        workspaces_to_remove = [workspace for workspace in self.plotted_workspaces if workspace not in workspace_list]
+        for workspace in workspaces_to_remove:
```
Comment (TTitcombe) [inline scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_model.py]: can we change this to
```
for workspace in self.plotted_workspaces:
    if workspace not in workspace_list:
        self.remove_workspace_from_plot(workspace)
```
to remove the second loop

File: scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_presenter.py [hunk 3]
```diff
@@ -107,20 +108,32 @@ def plot_standard_workspaces(self):
 
         self._model.plot(workspace_list, self.get_plot_title())
 
+        workspace_list_inverse_binning = self.get_workspaces_to_plot(self.context.group_pair_context.selected,
+                                                                     not self._view.if_raw(),
+                                                                     self._view.get_selected())
+        self._model.plotted_workspaces_inverse_binning = workspace_list_inverse_binning
+
+        if self.context.fitting_context.fit_list and \
+                any([workspace in workspace_list + workspace_list_inverse_binning
```
Comment (TTitcombe) [inline scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_presenter.py]: Can we add a short comment describing what we're trying to achieve in this statement?
I think also having `combined_ws_list = workspace_list + workspace_list_inverse_binning` and then `workspace in combined_ws_list` would be a better solution, just so we don't do `workspace_list + workspace_list_inverse_binning` for every workspace in the loop

File: scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_presenter.py [hunk 4]
```diff
@@ -107,20 +108,32 @@ def plot_standard_workspaces(self):
...
         """
         When a new fit is done adds the fit to the plotted workspaces if appropriate
         :return:
         """
-        for plotted_workspace in self._model.plotted_workspaces:
-            list_of_workspaces_to_plot = self.context.fitting_context.find_output_workspaces_for_input_workspace_name(plotted_workspace)
+        current_fit = self.context.fitting_context.fit_list[-1]
+        list_of_output_workspaces_to_plot = [output for output, input in
+                                             zip(current_fit.output_workspace_names, current_fit.input_workspaces)
+                                             if input in self._model.plotted_workspaces + self._model.plotted_workspaces_inverse_binning]
```
Comment (TTitcombe) [inline scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_presenter.py]: same here, regarding not appending the lists for every item in the for loop

**Review by TTitcombe — CHANGES_REQUESTED** [review]

File: scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_model.py (hunk 2 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/home_plot_widget/home_plot_widget_model.py]: The remove_workspace_from_plot function modifies the self.plotted_workspaces list removing the relevant work-spaces from it via a list comprehension. Whilst it's possible to remove items from a list in one pass for the size of the lists we are iterating over here I don't think performance is enough of a concern to warrant the increase in complexity.

**Review by TTitcombe — APPROVED** [review]

**Comment by TTitcombe** [comment]
Works as expected during testing.
Appropriate unit tests to check that plots are not being rescaled

**Pull request merged**